x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/DerParser.java

```diff
@@ -0,0 +1,277 @@
+/*
+  Copyright (c) 1998-2010 AOL Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+ */
+
+package org.elasticsearch.xpack.core.ssl;
+
+
+import org.elasticsearch.common.hash.MessageDigests;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
+
+/**
+ * A bare-minimum ASN.1 DER decoder, just having enough functions to
+ * decode PKCS#1 private keys in order to remain JCE/JVM agnostic.
+ * <p>
+ * Based on https://github.com/groovenauts/jmeter_oauth_plugin/blob/master/jmeter/src/
+ * main/java/org/apache/jmeter/protocol/oauth/sampler/PrivateKeyReader.java
+ */
+class DerParser {
+    // Constructed Flag
+    private static final int CONSTRUCTED = 0x20;
+
+    // Tag and data types
+    private static final int INTEGER = 0x02;
+    private static final int OCTET_STRING = 0x04;
+    private static final int OBJECT_OID = 0x06;
+    private static final int NUMERIC_STRING = 0x12;
+    private static final int PRINTABLE_STRING = 0x13;
+    private static final int VIDEOTEX_STRING = 0x15;
+    private static final int IA5_STRING = 0x16;
+    private static final int GRAPHIC_STRING = 0x19;
+    private static final int ISO646_STRING = 0x1A;
+    private static final int GENERAL_STRING = 0x1B;
+
+    private static final int UTF8_STRING = 0x0C;
+    private static final int UNIVERSAL_STRING = 0x1C;
+    private static final int BMP_STRING = 0x1E;
+
+
+    private InputStream derInputStream;
+
+    DerParser(byte[] bytes) {
+        this.derInputStream = new ByteArrayInputStream(bytes);
+    }
+
+    Asn1Object readAsn1Object() throws IOException {
+        int tag = derInputStream.read();
+        if (tag == -1) {
+            throw new IOException("Invalid DER: stream too short, missing tag");
+        }
+        int length = getLength();
+        byte[] value = new byte[length];
```

Comment:
I know we don't want to modify this code too much, but I think we should place some reasonable limits on this length.
I believe the `getLength` method can return any 32 bit integer (positive or negative), and we probably shouldn't just try to allocate an array of that length.